File: integrations/lib/tctl/resources.go
/*
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package tctl

import (
	"encoding/json"
	"errors"
	"io"

	"github.com/ghodss/yaml"
	"github.com/gravitational/trace"
	kyaml "k8s.io/apimachinery/pkg/util/yaml"

	"github.com/gravitational/teleport/api/types"
)

func writeResourcesYAML(w io.Writer, resources []types.Resource) error {
	for i, resource := range resources {
		data, err := yaml.Marshal(resource)
		if err != nil {
			return trace.Wrap(err)
		}
		if _, err := w.Write(data); err != nil {
			return trace.Wrap(err)
		}
		if i != len(resources) {
			if _, err := io.WriteString(w, "\n---\n"); err != nil {
				return trace.Wrap(err)
			}
		}
	}
	return nil
}

func readResourcesYAMLOrJSON(r io.Reader) ([]types.Resource, error) {
	var resources []types.Resource
	decoder := kyaml.NewYAMLOrJSONDecoder(r, 32768)
	for {
		var res streamResource
		err := decoder.Decode(&res)
		if err != nil {
			if errors.Is(err, io.EOF) {
				break
			}
			return nil, trace.Wrap(err)
		}
		resources = append(resources, res.Resource)
	}
	return resources, nil
}

type streamResource struct{ types.Resource }

func (res *streamResource) UnmarshalJSON(raw []byte) error {
	var header types.ResourceHeader
	if err := json.Unmarshal(raw, &header); err != nil {
		return trace.Wrap(err)
	}

	var resource types.Resource
	switch header.Kind {
	case types.KindNode:
		switch header.Version {
		case types.V2:
			resource = &types.ServerV2{}
		default:
			return trace.BadParameter("unsupported resource version %s", header.Version)
		}
	case types.KindUser:
		switch header.Version {
		case types.V2:
			resource = &types.UserV2{}
		default:
			return trace.BadParameter("unsupported resource version %s", header.Version)
		}
	case types.KindRole:
		switch header.Version {
		case types.V4, types.V5, types.V6, types.V7, types.V8:
			resource = &types.RoleV6{}
		default:
			return trace.BadParameter("unsupported resource version %s", header.Version)
		}
	case types.KindCertAuthority:
		switch header.Version {
		case types.V2:
			resource = &types.CertAuthorityV2{}
		default:
			return trace.BadParameter("unsupported resource version %s", header.Version)
		}
	default:
		return trace.BadParameter("unsupported resource kind %s", header.Kind)
	}

	if err := json.Unmarshal(raw, resource); err != nil {
		return trace.Wrap(err)
	}

	res.Resource = resource
	return nil
}

File: web/packages/teleport/src/test/helpers/resources.ts
/**
 * Teleport
 * Copyright (C) 2025 Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { http, HttpResponse } from 'msw';

import {
  ResourcesResponse,
  UnifiedResource,
} from 'teleport/services/agents/types';
import { JsonObject } from 'teleport/types';

// Using the path without query string params included
const path = '/v1/webapi/sites/:clusterId/resources';

export const fetchUnifiedResourcesSuccess = (opts?: {
  // FIXME: the response type should be raw API payload, not `UnifiedResource`.
  // `UnifiedResource` is produced later by `resourceService.fetchUnifiedResources()`
  // via `makeUnifiedResource`.
  // Mixing these layers can cause some values to be missing after going through makeUnifiedResource.
  response?: ResourcesResponse<UnifiedResource> & { items: UnifiedResource[] };
  delayMs?: number;
  mockSearch?: boolean;
}) => {
  const {
    response = {
      startKey: 'next',
      totalCount: 12,
      items: [
        {
          kind: 'kube_cluster',
          name: 'kube-lon-dev-01.example.com',
          labels: [
            {
              name: 'env',
              value: 'dev',
            },
            {
              name: 'region',
              value: 'eu-west-1',
            },
            {
              name: 'foo',
              value: 'bar',
            },
            {
              name: 'hello',
              value: 'world',
            },
          ],
        },
        {
          kind: 'kube_cluster',
          name: 'kube-lon-prod-01.example.com',
          labels: [
            {
              name: 'env',
              value: 'prod',
            },
            {
              name: 'region',
              value: 'eu-west-1',
            },
          ],
        },
        {
          kind: 'kube_cluster',
          name: 'kube-lon-staging-01.example.com',
          labels: [
            {
              name: 'env',
              value: 'staging',
            },
            {
              name: 'region',
              value: 'eu-west-1',
            },
          ],
        },
        {
          kind: 'kube_cluster',
          name: 'kube-temp-01.example.com',
        },
      ],
    },
    delayMs = 0,
    mockSearch = false,
  } = opts ?? {};

  return http.get(path, async ({ request }) => {
    const url = new URL(request.url);
    const search = url.searchParams.get('search');

    if (delayMs) {
      await new Promise(res => setTimeout(res, delayMs));
    }

    return HttpResponse.json({
      ...response,
      ...(mockSearch && search
        ? {
            items: response.items.filter(
              a => a.kind != 'kube_cluster' || a.name.includes(search)
            ),
          }
        : {}),
    });
  });
};

export const fetchUnifiedResourcesError = (
  status: number,
  error: string | null = null,
  extras: JsonObject = {}
) =>
  http.get(path, () => {
    return HttpResponse.json(
      { error: { message: `${status} - ${error}` }, extras },
      { status }
    );
  });

export const fetchUnifiedResourcesForever = () =>
  http.get(
    path,
    () =>
      new Promise(() => {
        /* never resolved */
      })
  );
